tests/Unit/EmogrifierTest.php

```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+     *
+     * @return string[][]
+     */
+    private function getSelectorSpecificityCssDataSets(
+        $name,
+        $firstSelector,
+        $secondSelector,
+        $operator,
+        $matchedTagPart
+    ) {
+        return [
+            $name . ' (1st selector valid for test)' => $this->getMatchedCssDataSet($firstSelector, $matchedTagPart),
+            $name . ' (2nd selector valid for test)' => $this->getMatchedCssDataSet($secondSelector, $matchedTagPart),
+            // Check first selector matches when its rule is last unless the second has higher specificity
+            $name . ' (1st selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $secondSelector,
+                $firstSelector,
+                $operator === '<',
+                $matchedTagPart
+            ),
+            $name . ' (2nd selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $firstSelector,
+                $secondSelector,
+                $operator === '>',
+                $matchedTagPart
+            )
+        ];
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function selectorSpecificityCssDataProvider()
+    {
+        $selectorData = [
+            ['class more specific than types', '.p-1', 'html body p', '>', '<p class="p-1"'],
+            ['id more specific than class and types', '#p4', 'html body .p-4', '>', '<p class="p-4" id="p4"'],
+            [
+                'id more specific than 20 classes and types',
+                '#p4',
+                'html body ' . str_repeat('.p-4', 20),
+                '>',
+                '<p class="p-4" id="p4"'
+            ],
+            ['pseudo-class as specific as class', '.p-1', '*:first-child', '==', '<p class="p-1"'],
+            ['attribute as specific as class', 'span[title="bonjour"]', '.p-2 span', '==', '<span title="bonjour"'],
+            [':not alone does not increase specificity', 'p:not(* + *)', 'p', '==', '<p class="p-1"'],
+            [':not with type more specific than nothing', '.p-1:not(h1)', '.p-1', '>', '<p class="p-1"'],
+            [':not with type as specific as type', '*:not(h1)', 'p', '==', '<p class="p-1"'],
+            [
+                ':not with several types more specific than single type',
+                '*:not(html body h1)',
+                'p',
+                '>',
+                '<p class="p-1"'
+            ],
+            [':not with class more specific than types', '*:not(.p-2)', 'html body p', '>', '<p class="p-1"'],
+            [
+                ':not with class more specific than :not with types',
+                '*:not(.p-2)',
+                '*:not(html body h1)',
+                '>',
+                '<p class="p-1"'
+            ],
+            ['class more specific than :not with types', '.p-1', '*:not(html body h1)', '>', '<p class="p-1"'],
+        ];
+
+        $datasets = [];
+        foreach ($selectorData as $data) {
```

Comment:
I think we need different variable names here because currently both "data" and "selectorData" basically mean the same thing, but the names should make the difference clear.

Reply:
No longer applicable after refactoring